### PR TITLE
[CI:BUILD] Cirrus: Catch use of deprecated ioutils package

### PR DESCRIPTION
**Pull request by cevich**
At the time of this commit, there's no easier way to detect this using `golangci-lint` or the go tool (that I could find).

Signed-off-by: Chris Evich <cevich@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: contrib/cirrus/runner.sh
Comment (edsantiago) [inline contrib/cirrus/runner.sh:242]: I suspect that this test is going to trigger on this very PR :wink:

File: contrib/cirrus/runner.sh
Comment (edsantiago) [inline contrib/cirrus/runner.sh:238]: General comment: I find the logs from this script impossible to read. How about:
```bash
echo
echo "Checking for added ioutils"
```
...with a corresponding `else echo "Skipping check for ioutils"`? (the `else` is important because it differentiates "test passed" from "test never ran at all").

File: contrib/cirrus/runner.sh
Comment (edsantiago) [reply contrib/cirrus/runner.sh:242]: Update: it's going to hang, actually, because there's no argument to grep. Anyhow, this needs to be fixed: to grep only `.go` files, to ignore `vendor`, and probably more. With time - not today.

File: contrib/cirrus/runner.sh
Comment (edsantiago) [inline contrib/cirrus/runner.sh:244]: echoing the full diffs is going to be rather daunting on a revendor or other big PR

**Review by edsantiago — CHANGES_REQUESTED** [review]

**Comment by edsantiago** [comment]
Oh yeah. Can't actually submit any PRs until #15890 merges

File: contrib/cirrus/runner.sh
Comment (cevich) [reply contrib/cirrus/runner.sh:239]: good idea.

File: contrib/cirrus/runner.sh
Comment (cevich) [reply contrib/cirrus/runner.sh:244]: damn.

**Comment by cevich** [comment]
> Oh yeah. Can't actually submit any PRs until https://github.com/containers/podman/pull/15890 merges

Yep :disappointed:

File: contrib/cirrus/runner.sh
Comment (cevich) [reply contrib/cirrus/runner.sh:244]: Actually we want it to fail on `.go` and `.vendor`, they will also fail to compile under go 1.19.

File: contrib/cirrus/runner.sh
Comment (cevich) [inline contrib/cirrus/runner.sh:240]: FYI- Cirrus logs squash empty lines.

File: contrib/cirrus/runner.sh
Comment (cevich) [inline contrib/cirrus/runner.sh:249]: This is imperfect, but simple.  Hopefully 10-lines of context is enough for a PR author to find their problem.

File: contrib/cirrus/runner.sh
Comment (Luap99) [inline contrib/cirrus/runner.sh:241]: The package is called `io/ioutil` without s. Please fix this also in the commit message, and the regex below

Comment (cevich) [reply contrib/cirrus/runner.sh:241]: Good catch, thanks.

File: contrib/cirrus/runner.sh
Comment (cevich) [inline contrib/cirrus/runner.sh:247]: Ugh, somehow [this is still matching line 246](https://cirrus-ci.com/task/5306032040706048) despite the obfuscation.

Comment (edsantiago) [reply contrib/cirrus/runner.sh:247]: No, it's matching the one two lines down.

I have strong concerns about this approach, but won't have time to fully review until later today.

Comment (cevich) [reply contrib/cirrus/runner.sh:247]: Oh, it's the messages.

**Comment by cevich** [comment]
Okay, this is [working now as intended](https://cirrus-ci.com/task/6249028416438272).

**Review by Luap99 — COMMENTED** [review]
LGTM

File: contrib/cirrus/runner.sh
Comment (edsantiago) [inline contrib/cirrus/runner.sh:242]: So....... you could just say `i=i` :wink:

Comment (cevich) [reply contrib/cirrus/runner.sh:242]: oh durrr, ya, I s'pose I could.  Thanks, I'll fix it.

**Comment by edsantiago** [comment]
Okay I'm not going to place a hold, but I think this PR is a very very bad idea. I think at an absolute minimum this needs to exclude `vendor`. It should also only diff `.go` files, so stuff like a Changelog entry mentioning i-o-utils will not be blocked. I've been reading git-diff help today in hopes of finding a way to solve the `.go` problem, but no luck so far.

**Comment by cevich** [comment]
> this needs to exclude vendor

No it must not, those bits are compiled also (golang 1.19 refuses to compile code that uses `ioutil` package).

Catching up non-go files is a legitimate concern, and something that would be nice to fix.  But the danger of somebody adding new `ioutil` code is pretty high, it's already happened at least twice while this PR has been open :frowning_face:

**Comment by cevich** [comment]
>  It should also only diff .go files

idea (for go-files fix): Just dump out changed filenames from the diff.  Loop over any matching `*.go`, checking the individual diffs for the ioutil regex.

**Comment by cevich** [comment]
Ideally, the best thing would be to wait for [this proposal](https://github.com/golang/go/issues/40364) to come through as suggested in https://stackoverflow.com/questions/71217127/how-to-detect-deprecated-direct-dependencies-in-a-pipeline-for-a-go-project

**Comment by Luap99** [comment]
> No it must not, those bits are compiled also (golang 1.19 refuses to compile code that uses `ioutil` package).

I doubt that this is true. Golang has a compatibility promise and I highly doubt that they just broke that. Are you sure it is not just the linter complaining?

**Comment by cevich** [comment]
> re you sure it is not just the linter complaining?

Oh it could be, you're probably right.

**Comment by edsantiago** [comment]
>Just dump out changed filenames from the diff. Loop over any matching *.go,

Can't: doesn't handle renames or removed files. (Without ugly muckery on the part of the script).

Here's how to do it:
```bash
git diff OLD NEW  -- '*.go' :^vendor/
```
Why didn't I go to stackoverflow in the first place? Why did I waste time (off and on, not all day, thankfully) reading git man pages?

**Comment by rhatdan** [comment]
While we are doing this we should also check for `os.IsNotExists(err)` versus the recommended  `errors.Is(err, os.ErrNotExist)`

**Comment by cevich** [comment]
> Why didn't I go to stackoverflow in the first place?

Yeah, I'm struggling with this too.  StackOverflow first, then Google, then `man` :rofl:

**Comment by cevich** [comment]
> Can't: doesn't handle renames or removed files.

We probably don't care about removed files, but renames for sure.  I'll try `git diff OLD NEW  -- '*.go' :^vendor/` but again, I'm worried about getting /something/ in place before more `ioutil` usage sneaks in and breaks my F37 PR.  So I don't want to spend too long on it.

> While we are doing this

Sure, but I would prefer to do that as a followup PR.  Everybody and their uncle has trained-muscle-memory for typing `ioutil...` :disappointed:

**Comment by cevich** [comment]
Force-push: Rebased + implemented Ed's suggestion.

**Comment by cevich** [comment]
Okay, [this seems to be working for a renamed file.](https://cirrus-ci.com/task/5959301515706368)

**Review by edsantiago — APPROVED** [review]

**Comment by edsantiago** [comment]
/lgtm
/hold

Thanks @cevich

**Pull request merged**
